File: Store/StoreFrontend/src/services/orderService.ts
import { Card } from '@/shared/interfaces/Card'
import { Order } from '@/shared/interfaces/Order'
import { OrderProduct } from '@/shared/interfaces/OrderProduct'
import { RequestCreateOrder } from '@/shared/interfaces/RequestCreateOrder'
import axios from 'axios'

const URL_order = process.env.NEXT_PUBLIC_HOST
	? `http://${process.env.NEXT_PUBLIC_HOST}:${process.env.NEXT_PUBLIC_PORT}/store/Order`
	: '/store/Order'

export const orderService = {
	async getOrders(telegramId: number, finished?: boolean) {
		try {
			const data: Order[] = await (
				await axios.get(URL_order, {
					params: { tg_id: telegramId, finished },
				})
			).data

			return data
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}

			return []
		}
	},
	async createOrder(newOrder: RequestCreateOrder, telegramId: number) {
		try {
			await axios.post(URL_order, newOrder, { params: { tg_id: telegramId } })
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}
		}
	},
	async getOrder(telegramId: number, orderId: number) {
		try {
			const data: Order = await (
				await axios.get(`${URL_order}/${orderId}`, {
					params: { tg_id: telegramId },
				})
			).data

			return data
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}

			return null
		}
	},
	async removeOrder(telegramId: number, orderId: number) {
		try {
			await axios.delete(`${URL_order}/${orderId}`, {
				params: { tg_id: telegramId },
			})
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}
		}
	},
	async createOrderProduct(
		telegramId: number,
		orderId: number,
		newOrderProduct: OrderProduct
	) {
		try {
			await axios.post(`${URL_order}/${orderId}/products`, newOrderProduct, {
				params: { tg_id: telegramId },
			})
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}
		}
	},
	async getOrderProducts(telegramId: number, orderId: number) {
		try {
			const data: OrderProduct[] = await (
				await axios.get(`${URL_order}/${orderId}/products`, {
					params: { tg_id: telegramId },
				})
			).data

			return data
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}

			return null
		}
	},
	async removeOrderProduct(
		telegramId: number,
		orderId: number,
		productId: number
	) {
		try {
			await axios.patch(
				`${URL_order}/${orderId}/product/${productId}/delete`,
				{},
				{ params: { tg_id: telegramId } }
			)
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}
		}
	},
	async confirmOrder(telegramId: number, orderId: number, sectionId: number) {
		try {
			await axios.patch(
				`${URL_order}/${orderId}/confirm`,
				{},
				{ params: { section_id: sectionId, tg_id: telegramId } }
			)
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}
		}
	},
	async getOrderCard(telegramId: number, orderId: number) {
		try {
			const data: Card = (
				await axios.get(`${URL_order}/${orderId}/card`, {
					params: { tg_id: telegramId },
				})
			).data

			return data
		} catch (error) {
			if (error instanceof Error) {
				console.log(error.message)
			}

			return null
		}
	},
}

File: Store/StoreTgBot/src/services/order_service.py
import logging

from .store_api import StoreApiService
from src.models import domain, order, card, comment
from typing import List, Optional
from pydantic.v1 import parse_obj_as
from pydantic.v1 import parse_raw_as
from json import loads


class OrderService(StoreApiService):
    __instance = None

    def __new__(cls, *args, **kwargs):
        if cls.__instance is None:
            cls.__instance = super().__new__(cls)

        return cls.__instance

    def __init__(self, api_key: str = None, logger: logging.Logger = None):
        super().__init__(api_key=api_key)
        self.logger = logger
        self.controller = "/order"

    async def get_order(self, tg_id: str, order_id: int) -> domain.Order:
        url = self.api_key + self.controller + f"/{order_id}"
        params = {
            "tg_id": tg_id
        }
        async with self.session.get(url, params=params) as response:
            if response.status == 200:
                data = await response.json()
                return domain.Order.parse_obj(data)

    async def create_order(self, tg_id: str) -> domain.Order:
        url = self.api_key + self.controller
        params = {
            "tg_id": tg_id
        }
        request = loads(order.RequestCreateOrder(user_id=0, card_id=0, finished=False, price=0).json())
        async with self.session.post(url, params=params, json=request) as response:
            if response.status == 200:
                data = await response.json()
                return domain.Order.parse_obj(data)

    async def orders(self, tg_id: str, finished: Optional[bool] = None) -> List[domain.Order]:
        url = self.api_key + self.controller
        params = {
            "tg_id": tg_id
        }
        if finished is not None:
            params["finished"] = "true" if finished else "false"
        self.logger.info(params)
        async with self.session.get(url, params=params) as response:
            if response.status == 200:
                data = await response.json()
                return parse_obj_as(List[domain.Order], data)

    async def delete_order(self, tg_id: str, order_id: int) -> bool:
        url = self.api_key + self.controller + f"/{order_id}"
        params = {
            "tg_id": tg_id
        }
        async with self.session.delete(url, params=params) as response:
            data = await response.read()
            return parse_raw_as(bool, data)

    async def order_products(self, tg_id: str, order_id: int) -> List[order.OrderProduct]:
        url = self.api_key + self.controller + f"/{order_id}/products"
        params = {
            "tg_id": tg_id
        }
        async with self.session.get(url, params=params) as response:
            data = await response.json()
            return parse_obj_as(List[order.OrderProduct], data)

    async def add_order_product(self, tg_id: str, order_id: int, product_id: int,
                                quantity: int = 1) -> order.OrderProduct:
        url = self.api_key + self.controller + f"/{order_id}/products"
        params = {
            "tg_id": tg_id
        }
        request = loads(
            order.OrderProduct(order_id=order_id, product_id=product_id, quantity=quantity).json(exclude_none=False))
        async with self.session.post(url, params=params, json=request) as response:
            data = await response.json()
            return order.OrderProduct.parse_obj(data)

    async def delete_order_product(self, tg_id: str, order_id: int, product_id: int) -> domain.Order:
        url = self.api_key + self.controller + f"/{order_id}/product/{product_id}/delete"
        params = {
            "tg_id": tg_id
        }
        async with self.session.patch(url, params=params) as response:
            data = await response.json()
            return domain.Order.parse_obj(data)

    async def confirm_order(self, tg_id: str, order_id: int) -> domain.Card:
        url = self.api_key + self.controller + f"/{order_id}/confirm"
        params = {
            "tg_id": tg_id
        }
        async with self.session.patch(url, params=params) as response:
            data = await response.json()
            self.logger.info(data)
            # return card.ResponseGetCardDto.parse_obj(data)
            return domain.Card.parse_obj(data)

    async def order_card(self, tg_id: str, order_id: int) -> card.ResponseGetCardDto:
        url = self.api_key + self.controller + f"/{order_id}/card"
        params = {
            "tg_id": tg_id
        }
        async with self.session.get(url, params=params) as response:
            data = await response.json()
            return card.ResponseGetCardDto.parse_obj(data)

    async def order_comments(self, order_id: int) -> List[domain.Comment]:
        url = self.api_key + self.controller + f"/{order_id}/comments"
        async with self.session.get(url) as response:
            data = await response.json()
            return parse_obj_as(List[domain.Comment], data)

    async def add_comment(self, order_id: str, description: str) -> None:
        url = self.api_key + self.controller + f"/{order_id}/comments"
        request = loads(
            comment.RequestCreateCommentDto(description=description).json(exclude_none=False))
        async with self.session.post(url=url, json=request) as response:
            data = await response.json()
            return None
